backtracking/subarray_sum.cpp

```diff
@@ -0,0 +1,117 @@
+/**
+ * @file
+ * @brief We are given with an array and a sum value. The algorithms find all
+ * the subarrays of that array with sum equal to given sum and return such
+ * subarrays count. This approach will have \f$O(n)\f$ time complexity and
+ * \f$O(n)\f$ space complexity. NOTE: In this problem, we are only refering to
+ * the continuous subsets as subarrays everywhere. Subarrays can be created
+ * using deletion operation at the end or the front of an array only. The parent
+ * array is also counted in subarrays having 0 number of deletion operations.
+ * @details Subset sum(only continuous subsets) problem
+ * (https://en.wikipedia.org/wiki/Subset_sum_problem)
+ * @author [Swastika Gupta](https://github.com/swastyy)
+ */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for io operations
+#include <unordered_map>  /// for unordered_map
+#include <vector>         /// for std::vector
+
+/**
+ * @namespace backtracking
+ * @brief subarray sum algorithm
+ */
+namespace backtracking {
+/**
+ * @namespace Subarrays
+ * @brief Functions for counting subsets(only continuous subarrays) in a given
+ * array with a given sum Time Complexity: O(n), where ‘n’ is the number of
+ * elements in the given array.
+ */
+namespace Subarrays {
+/**
+ * @brief The main function implements count of subarrays
+ * @param sum is the required sum of any subarrays
+ * @param in_arr is the input array
+ * @returns count of the number of subsets with required sum
+ */
+
+std::uint64_t subarray_sum(int sum, const std::vector<int> &in_arr) {
+    int nelement = in_arr.size();
```

Comment:
Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

Reply:
Since we are taking arrays with negative elements too either we can make a template <typename T> or int will work for both